Source/WTF/wtf/generic/MainThreadGeneric.cpp

```diff
@@ -32,32 +32,21 @@
 
 #include "config.h"
 #include <pthread.h>
-#if HAVE(PTHREAD_NP_H)
-#include <pthread_np.h>
-#endif
 
 #include <wtf/RunLoop.h>
 
 namespace WTF {
 
-#if !HAVE(PTHREAD_MAIN_NP)
 static pthread_t mainThread;
-#endif
 
 void initializeMainThreadPlatform()
 {
-#if !HAVE(PTHREAD_MAIN_NP)
     mainThread = pthread_self();
-#endif
 }
 
 bool isMainThread()
 {
-#if HAVE(PTHREAD_MAIN_NP)
-    return pthread_main_np();
-#else
     return pthread_equal(pthread_self(), mainThread);
-#endif
```

Comment:
Note that macOS uses `pthread_main_np()` here. I don't think that the title of this PR accurate - it may do what's necessary on Linux, but it doesn't align it with other platforms.

On macOS and iOS, WebKit2 APIs must be called from the main thread (same as UI thread, same as what pthread_main_np considers main). Similarly, there is a number of platforms APIs that are main thread only, so WebKit hops to the main thread to use them. So the Eclipse use case described here would be invalid for us.

Reply:
> I don't think that the title of this PR accurate - it may do what's necessary on Linux, but it doesn't align it with other platforms.

AFAIU it's not a Linux specific issue so much as an issue dependent on how one uses webkit(ie if one launches it via a non-process main thread with GTK).

> same as UI thread

I think this is still the case on Linux, AFAIU main thread is still always the UI thread, it's just that both may not be the process main thread in some cases.

> On macOS and iOS, WebKit2 APIs must be called from the main thread (same as UI thread, same as what pthread_main_np considers main). Similarly, there is a number of platforms APIs that are main thread only, so WebKit hops to the main thread to use them. So the Eclipse use case described here would be invalid for us.

Hmm, what API's are those? Maybe we need something with different semantics for those cases like a `isProcessMainThread()`, that way other platforms like Linux could also make use of that for cases where one actually wants to see if one is running on the main process thread as opposed to just the thread which initialized webkit.

Reply:
APIs in `Source/WebKit/UIProcess/API/Cocoa/` are currently only to be called from the main thread. So, initialization always happens on the main thread.

It's different from JavaScriptCore, which is usable from secondary threads.

Again, I'm _not_ saying that this patch is wrong. But it's not making semantics the same across all platforms.

Reply:
I'd say this change to MainThreadGeneric is surely good, but the commit message is indeed not right. You're unifying the semantics of isMainThread for non-Cocoa Unix platforms in a way that will fix applications that would currently be broken on non-Linux platforms, so that BSD users have a chance to use applications that depend on this. That's good. But the semantics for Cocoa are still different.

I wouldn't try to change the Cocoa implementation to match MainThreadGeneric; regardless of whether the change would be correct on Apple platforms or not, evidently no applications need it or they would have complained by now. So I would leave the code exactly as you've done in this pull request, and instead just come up with a different commit message. That way we don't need to figure out whether it's OK for the WebKit main thread to not be the same as the operating system main thread on Apple platforms. Maybe you could add some comments to explain things, though, since this is very subtle and would be easy to break again in the future. Your original attempt to change this seemed pretty reasonable, after all, and someone might well try again at some point if we don't leave a warning.

Reply:
> I'd say this change to MainThreadGeneric is surely good, but the commit message is indeed not right. You're unifying the semantics of isMainThread for non-Cocoa Unix platforms in a way that will fix applications that would currently be broken on non-Linux platforms, so that BSD users have a chance to use applications that depend on this. That's good. But the semantics for Cocoa are still different.

I reworded it a little, is that better?

> Maybe you could add some comments to explain things, though, since this is very subtle and would be easy to break again in the future. Your original attempt to change this seemed pretty reasonable, after all, and someone might well try again at some point if we don't leave a warning.

Added a comment so that it's clear that `mainThread` may not be the process initial thread.

Reply:
> I reworded it a little, is that better?

No, because the behavior is not the same across all operating systems. :P  I would say something like "[WTF] Remove pthread_main_np use in MainThreadGeneric."

Reply:
> No, because the behavior is not the same across all operating systems. :P

Well I was referring to the implementation behavior of just `MainThreadGeneric` itself(ie eliminating the differing codepaths for `HAVE(PTHREAD_MAIN_NP)`).

> I would say something like "[WTF] Remove pthread_main_np use in MainThreadGeneric."

Changed to that.